src/s3/checksum.c

```diff
@@ -0,0 +1,254 @@
+/*-------------------------------------------------------------------------
+ *
+ * checksum.c
+ * 		Declarations for calculating checksums of S3-specific data.
+ *
+ * Copyright (c) 2024, Oriole DB Inc.
+ *
+ * IDENTIFICATION
+ *	  contrib/orioledb/src/s3/checksum.c
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "orioledb.h"
+
+#include "s3/checksum.h"
+
+#include "common/string.h"
+#include "storage/fd.h"
+#include "utils/hsearch.h"
+
+#include "openssl/sha.h"
+
+static void initHashTable(S3ChecksumState *state, const char *filename);
+
+/*
+ * Allocate S3ChecksumState and initalize hashTable by reading hashFilename.  A
+ * caller should prepare a buffer for S3FileChecksum entries, the caller is
+ * responsible to release the buffer.
+ */
+S3ChecksumState *
+makeS3ChecksumState(uint32 checkpointNumber, S3FileChecksum *pgFiles,
+					uint32 pgFilesMaxLen, const char *hashFilename)
+{
+	S3ChecksumState *res;
+
+	res = (S3ChecksumState *) palloc(sizeof(S3ChecksumState));
+	res->hashTable = NULL;
+	res->checkpointNumber = checkpointNumber;
+	res->pgFiles = pgFiles;
+	res->pgFilesMaxLen = pgFilesMaxLen;
+	res->pgFilesLen = 0;
+
+	initHashTable(res, hashFilename);
+
+	return res;
+}
+
+/*
+ * Free S3ChecksumState and hashTable.  A caller is responsible to release the
+ * buffer of S3FileChecksum entries.
+ */
+void
+freeS3ChecksumState(S3ChecksumState *state)
+{
+	if (state == NULL)
+		return;
+
+	hash_destroy(state->hashTable);
+	pfree(state);
+}
+
+/*
+ * Initialize a hash table to store checksums of PGDATA files.
+ */
+static void
+initHashTable(S3ChecksumState *state, const char *filename)
```

Comment:
I think function not only inits but also inserts. Provided it's called only once, maybe make its code a part of makeS3ChecksumState.

Reply:
I think having `initHashTable()` as a separate function makes `makeS3ChecksumState()` a bit more clear and hides initialization details from it.
"init" here means not only that the hash table is created but also initialized with initial data.